draft-ietf-sframe-enc.md

```diff
@@ -311,28 +303,61 @@ Key or Key Length (K, 3 bits):
 : This field contains the key id (KID) if the X flag is set to 0, or the key
 length (KLEN) if set to 1.
 
-If X flag is 0, then the KID is in the range of 0-7 and the counter (CTR) is
-found in the next LEN bytes:
+Extended Counter Flag (Y, 1 bit):
+: Indicates if the key field contains the key id or the key length.
 
-~~~~~ aasvg
- 0 1 2 3 4 5 6 7
-+-+-----+-+-----+---------------------------------+
-|R|LEN  |0| KID |    CTR... (length=LEN)          |
-+-+-----+-+-----+---------------------------------+
-~~~~~
-{: title="SFrame header with short KID" }
+Counter or Counter Length (C, 3 bits):
+: This field contains the key id (KID) if the X flag is set to 0, or the key
+length (KLEN) if set to 1.
 
-If X flag is 1 then KLEN is the length of the key (KID) in bytes, minus one
-(the range of possible lengths is thus 1-8). The KID is encoded in
-the KLEN bytes following the metadata byte, and the counter (CTR) is encoded
-in the next LEN bytes:
+The Key ID and Counter fields are encoded as compact unsigned integers in
+network (big-endian) byte order.  If the value of one of these fields is in the
+range 0-7, then the value is carried in the corresponding bits of the config
+byte (K or C) and the corresponding flag (X or Y) is set to zero.  Otherwise,
+the value is MUST be encoded with the minimum number of bytes required and
+appended after the configuration byte, with the Key ID first and Counter second.
+The header field (K or C) is set to the number of bytes in the encoded value,
+minus one.  The value 000 represents a length of 1, 001 a length of 2, etc.
+This allows a 3-bit length field to represent the value lengths 1-8.
+
+The SFrame header can thus take one of the four forms shown in
+{{fig-sframe-header-cases}}, depending on which of the X and Y flags are set.
 
 ~~~~~ aasvg
- 0 1 2 3 4 5 6 7
-+-+-----+-+-----+---------------------------+---------------------------+
-|R|LEN  |1|KLEN |   KID... (length=KLEN)    |    CTR... (length=LEN)    |
-+-+-----+-+-----+---------------------------+---------------------------+
+KID < 8, CTR < 8:
++-+-----+-+-----+
+|0| KID |0| CTR |
++-+-----+-+-----+
+
+KID < 8, CTR >= 8:
++-+-----+-+-----+------------------------+
+|0| KID |1|CLEN |  CTR... (length=CLEN)  |
++-+-----+-+-----+------------------------+
+
+KID >= 8, CTR < 8:
++-+-----+-+-----+------------------------+
+|1|KLEN |0| CTR |  KID... (length=KLEN)  |
++-+-----+-+-----+------------------------+
+
+KID >= 8, CTR >= 8:
++-+-----+-+-----+------------------------+------------------------+
+|1|KLEN |1|CLEN |  KID... (length=KLEN)  |  CTR... (length=CLEN)  |
++-+-----+-+-----+------------------------+------------------------+
+~~~~~
+{: #fig-sframe-header-cases title="Forms of Encoded SFrame Header" }
+
+In ABNF {{!RFC5234}}, the form of the SFrame Header is represented by the rules
```

Comment:
Interim 2023-09-13:
* The ABNF is not useful, and also incorrect (`1*7` should be `1*8`)